rfd/0031-dynamic-app-db-registration.md

```diff
@@ -0,0 +1,355 @@
+and use `teleport.dev/origin` label to denote whether the app/database resource
+comes from static configuration or created dynamically. The rules are:
+
+- Users can't delete apps/databases with `teleport.dev/origin: config-file`
+  label.
+- Users can't set `teleport.dev/origin` label in their resources - it will be
+  auto-set to `dynamic` when they're created.
+- Users can't create apps/databases with the same name as one of existing
+  static apps/databases (to avoid various conflict resolution issues).
+
+**Multiple replicas**
+
+When multiple app/database agents pick up the same web app or database resource,
+they all add it to their proxying configuration. The connections that go to
+this app/database follow the same load-balancing rules which already exist
+today. This is no different than registering the same app/database multiple
+times via static configuration today.
+
+## Security implications
+
+In the current Teleport security model, the trust is established between
+components of the system i.e. when a new app or database service agent joins
+the cluster, it needs to present a valid auth token. Once the service has
+registered, updating its configuration to connect new web apps or databases
+does not require re-registering the service with the cluster.
+
+Dynamic app/database registration plugs into this model - when a new app or
+database resource is created, it is picked up by a service that is already
+part of the cluster. Managing app/database resources themselves is subject to
+RBAC checks so only users with appropriate permissions to `app_server` and
+`db_server` resources are allowed to create/remove them.
+
+## API changes
+
+Teleport API client currently provides the following methods for managing
+app/database resources (copied from [api/client/client.go](https://github.com/gravitational/teleport/blob/v7.0.0-dev.9/api/client/client.go)):
+
+```go
+GetAppServers(ctx context.Context, namespace string, skipValidation bool) ([]types.Server, error)
+UpsertAppServer(ctx context.Context, server types.Server) (*types.KeepAlive, error)
+DeleteAppServer(ctx context.Context, namespace, name string) error
+```
+
+```go
+GetDatabaseServers(ctx context.Context, namespace string, skipValidation bool) ([]types.DatabaseServer, error)
+UpsertDatabaseServer(ctx context.Context, server types.DatabaseServer) (*types.KeepAlive, error)
+DeleteDatabaseServer(ctx context.Context, namespace, hostID, name string) error
+```
+
+These methods were designed for internal usage by Teleport components e.g. when
+app/database agent registers with the cluster or for RBAC checks. As such, they
+are quite awkward for API client users. Changes to make them more API friendly:
+
+* If we decide to settle on `app` and `db` terminology (vs current `app_server`
+  and `db_server`), name them appropriately like `GetApps` or `DeleteDatabase`.
+* Switch to `Create` and `Update` methods instead of `Upsert` anti-pattern.
```

Comment:
What is the plan with `upsert` in general? I agree that it's an anti pattern. Do we want to maintain `upsert` in the backend, but provide `Create` and `Update` in the client/server? It would be good to know the plan before I continue with #6394 as I've heard/seen some mixed answers.

Reply:
I think that for new APIs we should use Create/Update everywhere incl. on the backend.

Reply:
At some point, we should write up an API conventions doc.
I wasn't aware that upsert was an anti-pattern either.